gtk4/src/subclass/im_context.rs

```diff
@@ -0,0 +1,512 @@
+        return true.to_glib();
+    }
+    false.to_glib()
+}
+
+unsafe extern "C" fn im_context_preedit_changed<T: IMContextImpl>(ptr: *mut ffi::GtkIMContext) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+
+    imp.preedit_changed(wrap.unsafe_cast_ref())
+}
+
+unsafe extern "C" fn im_context_preedit_end<T: IMContextImpl>(ptr: *mut ffi::GtkIMContext) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+
+    imp.preedit_end(wrap.unsafe_cast_ref())
+}
+
+unsafe extern "C" fn im_context_preedit_start<T: IMContextImpl>(ptr: *mut ffi::GtkIMContext) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+
+    imp.preedit_start(wrap.unsafe_cast_ref())
+}
+
+unsafe extern "C" fn im_context_reset<T: IMContextImpl>(ptr: *mut ffi::GtkIMContext) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+
+    imp.reset(wrap.unsafe_cast_ref())
+}
+
+unsafe extern "C" fn im_context_retrieve_surrounding<T: IMContextImpl>(
+    ptr: *mut ffi::GtkIMContext,
+) -> glib::ffi::gboolean {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+
+    imp.retrieve_surrounding(wrap.unsafe_cast_ref()).to_glib()
+}
+
+unsafe extern "C" fn im_context_set_client_widget<T: IMContextImpl>(
+    ptr: *mut ffi::GtkIMContext,
+    widgetptr: *mut ffi::GtkWidget,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+    let widget: Borrowed<Option<Widget>> = from_glib_borrow(widgetptr);
+
+    imp.set_client_widget(wrap.unsafe_cast_ref(), widget.as_ref().as_ref());
+}
+
+unsafe extern "C" fn im_context_set_cursor_location<T: IMContextImpl>(
+    ptr: *mut ffi::GtkIMContext,
+    areaptr: *mut gdk::ffi::GdkRectangle,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+    let mut area = gdk::Rectangle::uninitialized();
+
+    imp.set_cursor_location(wrap.unsafe_cast_ref(), &mut area);
+    *areaptr = *area.to_glib_none().0;
+}
+
+unsafe extern "C" fn im_context_set_surrounding<T: IMContextImpl>(
+    ptr: *mut ffi::GtkIMContext,
+    textptr: *const libc::c_char,
+    length: i32,
+    cursor_index: i32,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+    let text: Borrowed<GString> = from_glib_borrow(textptr);
+
+    imp.set_surrounding(
+        wrap.unsafe_cast_ref(),
+        &text[0..(length as usize)],
```

Comment:
> the length of text , or -1 if text is nul-terminated

So you need to handle `-1` here differently. `-1 == usize::MAX` would panic here now :)